resource-service/src/main/java/uk/gov/gchq/palisade/service/resource/service/FunctionalIterator.java

```diff
@@ -0,0 +1,313 @@
+/*
+ * Copyright 2020 Crown Copyright
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package uk.gov.gchq.palisade.service.resource.service;
+
+import java.util.Iterator;
+import java.util.NoSuchElementException;
+import java.util.function.Consumer;
+import java.util.function.Function;
+import java.util.function.Predicate;
+import java.util.function.UnaryOperator;
+
+/**
+ * An interface that allows an {@link Iterator} to have Stream like methods.
+ *
+ * @param <T> the type of the {@link FunctionalIterator}
+ */
+public interface FunctionalIterator<T> extends Iterator<T> {
```

Comment:
I left a comment in readers saying i was unsure if this should be in common or not.
Does this class need to be duplicated or should we just use the instance of it that is in readers (in the hadoop-resource-service package i presume, which we have imported).... ?
Not sure what the best choice here is

Reply:
I think the best choice, for now, is to leave it in the 2 places and there can be a discussion about what to do for the resource-service implementations. The problem is that the FunctionalIterator in the readers repo is in the hadoop-readers module, not the hadoop-resource module.

Reply:
Left it in for now but we can have a discussion about this in the future.